api/src/functions/govDelivery/app.ts

```diff
@@ -0,0 +1,75 @@
+import { addNewsletterGovDeliveryFactory } from "src/domain/newsletter/addNewsletterGovDeliveryFactory";
+import { shouldAddToNewsletter } from "src/api/userRouter";
+import { LogWriter } from "@libs/logWriter";
+
+export default async function handler() {
+  const IS_OFFLINE = process.env.IS_OFFLINE === "true" || false; // set by serverless-offline
+  const IS_DOCKER = process.env.IS_DOCKER === "true" || false; // set in docker-compose
+  const USERS_TABLE = process.env.USERS_TABLE || "users-table-local";
+  const STAGE = process.env.STAGE || "local";
+
+  const DYNAMO_OFFLINE_PORT = process.env.DYNAMO_PORT || 8000;
+  let dynamoDb: AWS.DynamoDB;
+  let dynamoDbDocument: AWS.DynamoDB.DocumentClient;
+  if (IS_OFFLINE) {
+    let dynamoDbEndpoint = "localhost";
+    if (IS_DOCKER) {
+      dynamoDbEndpoint = "dynamodb-local";
+    }
+    const endpoint = {
+      region: "localhost",
+      endpoint: `http://${dynamoDbEndpoint}:${DYNAMO_OFFLINE_PORT}`,
+    };
+    dynamoDb = new AWS.DynamoDB(endpoint);
+    dynamoDbDocument = new AWS.DynamoDB.DocumentClient(endpoint);
+  } else {
+    dynamoDb = new AWS.DynamoDB();
+    dynamoDbDocument = new AWS.DynamoDB.DocumentClient();
+  }
+  const logger = LogWriter(`NavigatorWebService/${STAGE}`, "SearchApis");
+  const dbClient = DynamoUserDataClient(dynamoDbDocument, USERS_TABLE);
+
+  const statement = `SELECT data FROM "${USERS_TABLE}" WHERE data["user"].receiveNewsletter = true and (data["user"].externalStatus.receiveNewsletter is missing or data["user"].externalStatus.receiveNewsletter.success = false)`;
+
+  const { Items = [] } = await dynamoDb.executeStatement({ Statement: statement }).promise();
+  const results = Items.map((i) => AWS.DynamoDB.Converter.unmarshall(i).data) as UserData[];
+  const GOV_DELIVERY_BASE_URL =
+    process.env.GOV_DELIVERY_BASE_URL ||
+    (IS_OFFLINE ? `http://${IS_DOCKER ? "wiremock" : "localhost"}:9000` : "https://api.govdelivery.com");
+  const GOV_DELIVERY_API_KEY = process.env.GOV_DELIVERY_API_KEY || "tempkey";
+  const GOV_DELIVERY_TOPIC = process.env.GOV_DELIVERY_TOPIC || "NJGOV_1";
+  const GOV_DELIVERY_URL_QUESTION_ID = process.env.GOV_DELIVERY_URL_QUESTION_ID || "q_86783";
+  const newsletterGovDeliveryClient = WebserviceNewsletterGovDeliveryClient(
+    GOV_DELIVERY_BASE_URL,
+    GOV_DELIVERY_TOPIC,
+    GOV_DELIVERY_API_KEY,
+    logger,
+    "navigator.business.nj.gov",
+    GOV_DELIVERY_URL_QUESTION_ID
+  );
+  const addNewsletterGovDelivery = addNewsletterGovDeliveryFactory(newsletterGovDeliveryClient);
+  let attempts = 0;
+  results.forEach(async (newUserData: UserData) => {
+    let userData = { ...newUserData };
+    if (shouldAddToNewsletter(userData)) {
+      attempts += 1;
+      userData = {
+        ...userData,
+        user: {
+          ...userData.user,
+          externalStatus: {
+            ...userData.user.externalStatus,
+            ...(await addNewsletterGovDelivery(userData.user.email)),
```

Comment:
similar question about await - I don't remember this but if one fails will it fail the whole foreach loop?

Reply:
this whole logic might also want to be a separate function